### ISSUE TITLE
Control logging frequency of Tensorboard Handlers

### ISSUE THREAD

**Issue by danieltudosiu** [Project-MONAI/MONAI#1549]
**Is your feature request related to a problem? Please describe.**
Saving the logs at the end of each iteration and epoch is not optimal for every task.

**Describe the solution you'd like**
Allow the user to pass the frequency of iterations and epochs at which the logging happens. This feature is easily implemented since Ignite already allows it.

**Additional context**
The lines of code that I refer to are here:
https://github.com/Project-MONAI/MONAI/blob/3cb82cfc781d8d67a1541076a6c5badc95f38978/monai/handlers/tensorboard_handlers.py#L89

**Comment by danieltudosiu** [comment]
I had a quick go at it and modified it resulting in this.


`

    class TensorBoardHandler(object):
        # TODO: Make it more feature rich and submit pull request to MONAI
        """
        TensorBoardStatsHandler defines a set of Ignite Event-handlers for all the TensorBoard logics.
        It's can be used for any Ignite Engine(trainer, validator and evaluator).
        And it can support both epoch level and iteration level with pre-defined TensorBoard event writer.
        The expected data source is Ignite ``engine.state.output`` and ``engine.state.metrics``.
        Default behaviors:
            - When EPOCH_COMPLETED, write each dictionary item in
              ``engine.state.metrics`` to TensorBoard.
            - When ITERATION_COMPLETED, write each dictionary item in
              ``self.output_transform(engine.state.output)`` to TensorBoard.
        """
    
        def __init__(
            self,
            summary_writer: Optional[SummaryWriter],
            log_dir: str = "./runs",
            epoch_interval: int = 1,
            epoch_event_writer: Optional[Callable[[Engine, SummaryWriter], Any]] = None,
            iteration_interval: int = 1,
            iteration_event_writer: Optional[Callable[[Engine, SummaryWriter], Any]] = None,
            output_transform: Callable = lambda x: x,
            global_epoch_transform: Callable = lambda x: x,
            tag_name: str = DEFAULT_TAG,
        ) -> None:
            """
            Args:
                summary_writer: user can specify TensorBoard SummaryWriter,
                    default to create a new writer.
                log_dir: if using default SummaryWriter, write logs to this directory, default is `./runs`.
                epoch_interval: plot content from engine.state.metrics every N epochs, default is 1.
                epoch_event_writer: customized callable TensorBoard writer for epoch level.
                    Must accept parameter "engine" and "summary_writer", use default event writer if None.
                iteration_interval: plot content from engine.state.output every N iterations, default is 1.
                iteration_event_writer: customized callable TensorBoard writer for iteration level.
                    Must accept parameter "engine" and "summary_writer", use default event writer if None.
                output_transform: a callable that is used to transform the
                    ``ignite.engine.output`` into a scalar to plot, or a dictionary of {key: scalar}.
                    In the latter case, the output string will be formatted as key: value.
                    By default this value plotting happens when every iteration completed.
                global_epoch_transform: a callable that is used to customize global epoch number.
                    For example, in evaluation, the evaluator engine might want to use trainer engines epoch number
                    when plotting epoch vs metric curves.
                tag_name: when iteration output is a scalar, tag_name is used to plot, defaults to ``'Loss'``.
            """
            self._writer = (
                SummaryWriter(log_dir=log_dir) if summary_writer is None else summary_writer
            )
            self.epoch_interval = epoch_interval
            self.epoch_event_writer = epoch_event_writer
            self.iteration_interval = iteration_interval
            self.iteration_event_writer = iteration_event_writer
            self.output_transform = output_transform
            self.global_epoch_transform = global_epoch_transform
            self.tag_name = tag_name
    
        def attach(self, engine: Engine) -> None:
            """
            Register a set of Ignite Event-Handlers to a specified Ignite engine.
    
            Args:
                engine: Ignite Engine, it can be a trainer, validator or evaluator.
    
            """
            if not engine.has_event_handler(
                self.iteration_completed, Events.ITERATION_COMPLETED
            ):
                engine.add_event_handler(
                    Events.ITERATION_COMPLETED(every=self.iteration_interval),
                    self.iteration_completed,
                )
            if not engine.has_event_handler(self.epoch_completed, Events.EPOCH_COMPLETED):
                engine.add_event_handler(
                    Events.EPOCH_COMPLETED(every=self.epoch_interval), self.epoch_completed
                )
    
        def epoch_completed(self, engine: Engine) -> None:
            """
            Handler for train or validation/evaluation epoch completed Event.
            Write epoch level events, default values are from Ignite state.metrics dict.
    
            Args:
                engine: Ignite Engine, it can be a trainer, validator or evaluator.
    
            """
            if self.epoch_event_writer is not None:
                self.epoch_event_writer(engine, self._writer)
            else:
                self._default_epoch_writer(engine, self._writer)
    
        def iteration_completed(self, engine: Engine) -> None:
            """
            Handler for train or validation/evaluation iteration completed Event.
            Write iteration level events, default values are from Ignite state.logs dict.
    
            Args:
                engine: Ignite Engine, it can be a trainer, validator or evaluator.
    
            """
            if self.iteration_event_writer is not None:
                self.iteration_event_writer(engine, self._writer)
            else:
                self._default_iteration_writer(engine, self._writer)
    
        def _default_epoch_writer(self, engine: Engine, writer: SummaryWriter) -> None:
            """
            Execute epoch level event write operation based on Ignite engine.state data.
            Default is to write the values from Ignite state.metrics dict.
    
            Args:
                engine: Ignite Engine, it can be a trainer, validator or evaluator.
                writer: TensorBoard writer, created in TensorBoardHandler.
    
            """
            current_epoch = self.global_epoch_transform(engine.state.epoch)
            summary_dict = engine.state.metrics
            for name, value in summary_dict.items():
                writer.add_scalar(name, value, current_epoch)
            writer.flush()
    
        def _default_iteration_writer(self, engine: Engine, writer: SummaryWriter) -> None:
            """
            Execute iteration level event write operation based on Ignite engine.state data.
            Default is to write the loss value of current iteration.
    
            Args:
                engine: Ignite Engine, it can be a trainer, validator or evaluator.
                writer: TensorBoard writer, created in TensorBoardHandler.
    
            """
            loss = self.output_transform(engine.state.output)
            if loss is None:
                return  # do nothing if output is empty
            if isinstance(loss, dict):
                for name in sorted(loss):
                    value = loss[name]
                    if not is_scalar(value):
                        warnings.warn(
                            "ignoring non-scalar output in TensorBoardStatsHandler,"
                            " make sure `output_transform(engine.state.output)` returns"
                            " a scalar or dictionary of key and scalar pairs to avoid this warning."
                            " {}:{}".format(name, type(value))
                        )
                        continue  # not plot multi dimensional output
                    writer.add_scalar(
                        name,
                        value.item() if torch.is_tensor(value) else value,
                        engine.state.iteration,
                    )
            elif is_scalar(loss):  # not printing multi dimensional output
                writer.add_scalar(
                    self.tag_name,
                    loss.item() if torch.is_tensor(loss) else loss,
                    engine.state.iteration,
                )
            else:
                warnings.warn(
                    "ignoring non-scalar output in TensorBoardStatsHandler,"
                    " make sure `output_transform(engine.state.output)` returns"
                    " a scalar or a dictionary of key and scalar pairs to avoid this warning."
                    " {}".format(type(loss))
                )
            writer.flush()
`

**Comment by wyli** [comment]
this makes sense, the option name could be `every` following https://github.com/pytorch/ignite/blob/702d3a2eed412118651004d6c7c03739ac33dfd4/ignite/engine/events.py#L63 do you have time to make a pull request?

**Comment by danieltudosiu** [comment]
Hi @wyli, If you are ok with it, I would like to give it a go and refractor both Tensorboard Handlers in a single one to reduce the code used and make it more intuitive. I hope to have a MVP this weekend.

**Comment by wyli** [comment]
thanks, we have some code duplications here and there on purpose...let's not rush refactoring them. For now decoupling the actual computing logic from the event handlers has a higher priority :)

**Comment by danieltudosiu** [comment]
Ok, will do just that and open another issue for merger of the two
handlers.

**Comment by danieltudosiu** [comment]
I did not forget about this feature request. Will create a PR for it soon.

**Comment by wyli** [comment]
ping

**Comment by danieltudosiu** [comment]
Sorry, you will have it later today.  I should branch from master right?

**Comment by danieltudosiu** [comment]
Hi @wyli 

I have a draft form the pull request, but I have the following questions:

- Should we change the step count from epoch to iteration and allow globa_step_transformation for the iteration level scalar summaries? This will normalize the step between the two levels of logging and will be more user-friendly since no mind maths needs to be done to sync the two loggings.
- How would you test the interval setting? Is there a way to count the times an event fires in Igntie?

Cheers!

**Comment by wyli** [comment]
> Hi @wyli
> 
> I have a draft form the pull request, but I have the following questions:
> 
> * Should we change the step count from epoch to iteration and allow globa_step_transformation for the iteration level scalar summaries? This will normalize the step between the two levels of logging and will be more user-friendly since no mind maths needs to be done to sync the two loggings.
> * How would you test the interval setting? Is there a way to count the times an event fires in Igntie?
> 
> Cheers!

thanks, could you please submit a draft pull request, it might be easier to discuss over the concrete code examples cc @Nic-Ma

**Comment by danieltudosiu** [comment]
I did it, you can find it here: https://github.com/Project-MONAI/MONAI/pull/1870

**Comment by wyli** [comment]
addressed by #1870
